projects/bootloader/scripts/test_can_datagram.py

```diff
@@ -0,0 +1,145 @@
+from can_datagram import DatagramSender
+from can_datagram import DatagramListener
+
+TEST_CHANNEL = "vcan0"
+
+TEST_PROTOCOL_VERSION = 1
+TEST_DATAGRAM_TYPE_ID = 1
+TEST_NODES = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+TEST_DATA = [3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5, 8, 9, 7, 9, 3, 2, 3, 8, 4, 6, 2, 6, 4, 3, 3]
+
+
+class TestCanDatagram(unittest.TestCase):
+    """Test Can Datagram functions"""
+
+    def test_create_message(self):
+        """Test the constructor of the Datagram class"""
+
+        message = Datagram(
+            datagram_type_id=TEST_DATAGRAM_TYPE_ID,
+            node_ids=TEST_NODES,
+            data=bytearray(TEST_DATA))
+
+        self.assertEqual(message._protocol_version, TEST_PROTOCOL_VERSION)
+        self.assertEqual(message._datagram_type_id, TEST_DATAGRAM_TYPE_ID)
+        self.assertEqual(message._node_ids, TEST_NODES)
+        self.assertEqual(message._data, bytearray(TEST_DATA))
+
+    def test_modify_message(self):
+        """Test modifying values in the Datagram class"""
+        message = Datagram(
+            protocol_version=TEST_PROTOCOL_VERSION,
+            datagram_type_id=TEST_DATAGRAM_TYPE_ID,
+            node_ids=TEST_NODES,
+            data=bytearray(TEST_DATA))
+
+        self.assertEqual(message._protocol_version, TEST_PROTOCOL_VERSION)
+        self.assertEqual(message._datagram_type_id, TEST_DATAGRAM_TYPE_ID)
+        self.assertEqual(message._node_ids, TEST_NODES)
+        self.assertEqual(message._data, bytearray(TEST_DATA))
+
+        protocol_version = 9
+        datagram_type_id = 8
+        test_nodes = list(reversed(TEST_NODES))
+        test_data = list(reversed(TEST_DATA))
+
+        message.protocol_version = protocol_version
+        message.datagram_type_id = datagram_type_id
+        message.node_ids = test_nodes
+        message.data = bytearray(test_data)
+
+        self.assertEqual(message._protocol_version, protocol_version)
+        self.assertEqual(message._datagram_type_id, datagram_type_id)
+        self.assertEqual(message._node_ids, test_nodes)
+        self.assertEqual(message._data, bytearray(test_data))
+
+    def test_serialize(self):
+        """Test the serialize function"""
+        message = Datagram(
+            protocol_version=TEST_PROTOCOL_VERSION,
+            datagram_type_id=TEST_DATAGRAM_TYPE_ID,
+            node_ids=TEST_NODES,
+            data=bytearray(TEST_DATA))
+
+        self.assertEqual(message.serialize(), bytearray(
+            b'\x01m\xc2\x18\xfe\x01\n\x01\x02\x03\x04\x05\x06\x07\x08\t\n\x1a\x00\x03\x01\x04\x01\x05\t\x02\x06\x05\x03\x05\x08\t\x07\t\x03\x02\x03\x08\x04\x06\x02\x06\x04\x03\x03'))
+
+    def test_deserialize(self):
+        """Test retrieving Datagram information from the bytearray"""
+        message = Datagram.deserialize(bytearray(
+            b'\x01m\xc2\x18\xfe\x01\n\x01\x02\x03\x04\x05\x06\x07\x08\t\n\x1a\x00\x03\x01\x04\x01\x05\t\x02\x06\x05\x03\x05\x08\t\x07\t\x03\x02\x03\x08\x04\x06\x02\x06\x04\x03\x03'))
```

Comment:
Did you generate these using this script? If so, it's probably better to generate them manually, to ensure that they're in adherence with the spec. You can define a bytearray (shared between the two tests) from an array of bytes like `bytearray([0x01, 0x20, ...])` and then add comments so someone can easily verify that it's in accordance with the spec. Then we can make sure that serialize/deserialize is strictly in accordance with the spec.